src/full/Agda/Utils/DocTree.hs

```diff
@@ -0,0 +1,156 @@
+-- | A simple document tree to render @'Doc' ann@ but preserve annotations.
+--
+-- 'DocTree' and 'renderToTree'' originally taken from
+-- <https://github.com/plt-amy/agda/blob/9fd50b883f14a05792ed79a0b693fbecb2165bf5/src/full/Agda/LSP/Output.hs>
+-- but rewritten to encode more invariants.
+
+module Agda.Utils.DocTree
+  ( DocTree
+  , treeToTextNoAnn
+  , treeToText
+  , renderTree'
+  , renderToTree
+  , renderToTree'
+  )
+where
+
+import Prelude hiding (null)
+
+import Data.Text          (Text)
+import Data.Text          qualified as Text
+-- -- TODO: not available on text-1.2.5.0 (GHC 9.2)
+-- import Data.Text.Encoding (StrictTextBuilder, strictBuilderToText, textToStrictBuilder)
+-- import Data.Text.Encoding qualified as Text
+
+import GHC.Generics
+
+import Text.PrettyPrint.Annotated.HughesPJ (Doc)
+import Text.PrettyPrint.Annotated.HughesPJ qualified as Ppr
+
+import Agda.Utils.List1 (List1, pattern (:|), (<|))
+import Agda.Utils.List1 qualified as List1
+import Agda.Utils.Null
+
+import Agda.Utils.Impossible
+import Agda.Utils.Function (applyUnless)
+
+-- | A rendered document with annotations from type @ann@.
+data DocTree ann
+  = Node ann [DocTree ann]
+     -- ^ Stuff annotated by @ann@.
+  | Text Text
+     -- ^ Atom.
+  deriving Generic
```

Comment:
@plt-amy I took your DocTree and ported your `Doc -> DocTree` renderer to the generic `ann` case.
That worked, but once I had a closer look at the renderer I engineered its true structure using a stack of yet-open annotation blocks as auxiliary structure.  So I got also rid of the bad `Text.cons` which is O(n).

Do you have a suggestion for a text-builder to transform the `DocTree` into `Text`? I went for `StrictTextBuilder` but then this is too new (we use older `text` versions).  There is `Data.Text.Lazy.Builder` (also in older `text`), would that be an option?